### PR TITLE
Allow long pressing for changing the offset of a time entry

### LINKED ISSUE #2930: Allow long pressing for changing the offset of a time entry
Design specs can be found at #2878

When you long press a time entry, you should be able to drag it to shift its start/end times. This was already working in the prototypes developed for #2758, so it shouldn't be too hard to integrate in the actual app.

Since there's a lot of coupling, both the foundation and the iOS parts should be handled by this issue. Thorough testing is still expected, though.

⚠️ This should only be applicable to created time entries in the calendar view, so it doesn't work for calendar events! ⚠️

### PR DESCRIPTION
**Pull request by juanlaube**
## What's this?

Allows to drag/resize items in the calendar collection view.

### Relationships

Closes #2930

## How is it done?

It's done by adding another helper class that wraps some gesture recognizers that mimics the native Calendar app behavior.
A first long press gesture enters in a drag and drop mode that allows moving the items.
A second pan gesture allows to resize the items.
Finally, there's a tap gesture to leave this editing mode.

### Why not another way?

The same logic could have been added to the previous gesture helper, but that class is long enough and independent of this feature.

### Side effects

None.

## Review hints

I recommend going commit by commit. The big stuff happens in https://github.com/toggl/mobileapp/pull/2993/commits/36184e9e133116cbf3bf9c7e8c863030f71d0af1 and https://github.com/toggl/mobileapp/pull/2993/commits/64c7e515f7f277ccf5d14f18c88fbf35b0ab5b34, the rest is just design, renaming or small stuff.

## :squid: Permissions

Myself.

### CODE REVIEW

File: Toggl.Daneel/Cells/Calendar/CalendarItemView.cs [hunk 1]
```diff
@@ -10,13 +12,22 @@
 
 namespace Toggl.Daneel.Cells.Calendar
 {
-    public sealed partial class CalendarItemView : ReactiveCollectionViewCell<CalendarItem>
+    public sealed partial class CalendarItemView : ReactiveCollectionViewCell<(CalendarItem, bool)>
```
Comment (willsb) [inline Toggl.Daneel/Cells/Calendar/CalendarItemView.cs]: No me gusta. Create a type that properly describe this tuple (a good candidate for name would be `SelectableCalendarViewModel`)

File: Toggl.Daneel/Cells/Calendar/CalendarItemView.cs [hunk 2]
```diff
@@ -40,21 +51,30 @@ public CalendarItemView(IntPtr handle) : base(handle)
...
+            topDragIndicatorBorderLayer.BorderWidth = 2;
+            topDragIndicatorBorderLayer.FillColor = UIColor.Clear.CGColor;
+            TopDragIndicator.Layer.AddSublayer(topDragIndicatorBorderLayer);
+
+            var bottomRect = TopDragIndicator.Bounds.Inset(1, 1);
+            bottomDragIndicatorBorderLayer = new CAShapeLayer();
+            bottomDragIndicatorBorderLayer.Path = UIBezierPath.FromOval(bottomRect).CGPath;
+            bottomDragIndicatorBorderLayer.BorderWidth = 2;
+            bottomDragIndicatorBorderLayer.FillColor = UIColor.Clear.CGColor;
+            BottomDragIndicator.Layer.AddSublayer(bottomDragIndicatorBorderLayer);
```
Comment (willsb) [inline Toggl.Daneel/Cells/Calendar/CalendarItemView.cs]: Repeated code could be a method

File: Toggl.Daneel/ViewSources/CalendarCollectionViewSource.cs [hunk 3]
```diff
@@ -78,7 +81,8 @@ public sealed class CalendarCollectionViewSource : MvxCollectionViewSource, ICal
         protected override UICollectionViewCell GetOrCreateCellFor(UICollectionView collectionView, NSIndexPath indexPath, object item)
         {
             var cell = collectionView.DequeueReusableCell(itemReuseIdentifier, indexPath) as CalendarItemView;
-            cell.Item = calendarItems[(int)indexPath.Item];
+            var isEditingItem = IsEditing && indexPath == editingItemIndexPath;
+            cell.Item = (calendarItems[(int)indexPath.Item], isEditingItem);
```
Comment (willsb) [inline Toggl.Daneel/ViewSources/CalendarCollectionViewSource.cs]: This is really cryptic, but now that I think about it maybe the type exposed from the VM is not what you want, but rather have a `cell.IsInEditMode` property instead.

File: Toggl.Daneel/Views/Calendar/CalendarCollectionViewEditItemHelper.cs [hunk 4]
```diff
@@ -0,0 +1,257 @@
...
+            {
+                case UIGestureRecognizerState.Began:
+                    longPressBegan(point);
+                    break;
+
+                case UIGestureRecognizerState.Changed:
+                    longPressChanged(point);
+                    break;
+
+                default:
```
Comment (willsb) [inline Toggl.Daneel/Views/Calendar/CalendarCollectionViewEditItemHelper.cs]: This default case is useless

File: Toggl.Daneel/Views/Calendar/CalendarCollectionViewEditItemHelper.cs [hunk 5]
```diff
@@ -0,0 +1,257 @@
...
+            {
+                case UIGestureRecognizerState.Began:
+                    panBegan(point);
+                    break;
+
+                case UIGestureRecognizerState.Changed:
+                    panChanged(point);
+                    break;
+
+                default:
```
Comment (willsb) [inline Toggl.Daneel/Views/Calendar/CalendarCollectionViewEditItemHelper.cs]: This default case is useless

File: Toggl.Daneel/Views/Calendar/CalendarCollectionViewEditItemHelper.cs [hunk 6]
```diff
@@ -0,0 +1,257 @@
...
+        {
+            if (dataSource.IsEditing && !isActive)
+                return;
+
+            calendarItem = dataSource.CalendarItemAtPoint(point).Value;
+            if (!calendarItem.Editable)
+                return;
+
+            itemIndexPath = collectionView.IndexPathForItemAtPoint(point);
+            dataSource.StartEditing(itemIndexPath);
```
Comment (willsb) [inline Toggl.Daneel/Views/Calendar/CalendarCollectionViewEditItemHelper.cs]: How does this ensure that we can't ever have more than one item being edited at any given time?

File: Toggl.Daneel/Views/Calendar/CalendarCollectionViewEditItemHelper.cs [hunk 7]
```diff
@@ -0,0 +1,257 @@
...
+        private void onTap(UITapGestureRecognizer gesture)
+        {
+            var point = gesture.LocationInView(collectionView);
+
+            if (collectionView.IndexPathForItemAtPoint(point) == itemIndexPath)
+                return;
+
+            isActive = false;
+            dataSource.StopEditing();
+            collectionView.RemoveGestureRecognizer(panGestureRecognizer);
```
Comment (willsb) [inline Toggl.Daneel/Views/Calendar/CalendarCollectionViewEditItemHelper.cs]: Maybe ensure that settings active to true _always_ adds the gesture recognizers and that setting it to false _always_  remove the same recognizers

File: Toggl.Foundation.MvvmCross/ViewModels/Calendar/CalendarViewModel.cs [hunk 8]
```diff
@@ -170,6 +174,28 @@ private IObservable<Unit> onDurationSelected(DateTimeOffset startTime, TimeSpan
                 await fetchCalendarItems(timeService.CurrentDateTime.Date);
             });
 
+        private IObservable<Unit> onUpdateTimeEntry(CalendarItem calendarItem)
+            => Observable.FromAsync(async () =>
+            {
+                var timeEntry = await dataSource.TimeEntries.GetById((calendarItem.TimeEntryId.Value));
```
Comment (willsb) [inline Toggl.Foundation.MvvmCross/ViewModels/Calendar/CalendarViewModel.cs]:
```csharp
if (!calendarItem.IsEditable() || calendarItem.TimeEntryId == null)
    return;
```

File: Toggl.Foundation.MvvmCross/ViewModels/Calendar/CalendarViewModel.cs [hunk 9]
```diff
@@ -170,6 +174,28 @@ private IObservable<Unit> onDurationSelected(DateTimeOffset startTime, TimeSpan
...
+
+                var dto = new DTOs.EditTimeEntryDto
```
Comment (willsb) [inline Toggl.Foundation.MvvmCross/ViewModels/Calendar/CalendarViewModel.cs]: Is this still the de facto way of updating TEs? There's nothing that uses `New<T>` as of yet?

File: Toggl.Foundation/Calendar/CalendarItem.cs [hunk 10]
```diff
@@ -25,6 +25,8 @@ public struct CalendarItem
 
         public CalendarIconKind IconKind { get; }
 
+        public bool Editable { get; }
```
Comment (willsb) [inline Toggl.Foundation/Calendar/CalendarItem.cs]: I don't like this. Make this an extension method like this:

```csharp
public static bool IsEditable(this CalendarItem calendarItem)
    => calendarItem.Source == CalendarItemSource.TimeEntry;
```

File: Toggl.Foundation/Calendar/CalendarItem.cs [hunk 11]
```diff
@@ -59,9 +63,14 @@ private CalendarItem(IThreadSafeTimeEntry timeEntry)
             {
                 case SyncStatus.SyncNeeded:
                     IconKind = CalendarIconKind.Unsynced;
+                    Editable = false;
                     break;
                 case SyncStatus.SyncFailed:
                     IconKind = CalendarIconKind.Unsyncable;
+                    Editable = false;
```
Comment (willsb) [inline Toggl.Foundation/Calendar/CalendarItem.cs]: You can (and should be able to!) edit unsyncable and unsynced time entries

File: Toggl.Foundation/Interactors/InteractorFactory.TimeEntry.cs [hunk 12]
```diff
@@ -49,10 +49,16 @@ public IInteractor<IObservable<IThreadSafeTimeEntry>> ContinueMostRecentTimeEntr
                 timeService,
                 dataSource,
                 analyticsService);
-        
+
+        public IInteractor<IObservable<IThreadSafeTimeEntry>> UpdateTimeEntry(DTOs.EditTimeEntryDto dto)
+            => new UpdateTimeEntryInteractor(
```
Comment (willsb) [inline Toggl.Foundation/Interactors/InteractorFactory.TimeEntry.cs]: There was no interactor previously? This is really weird. What did the edit view use?

File: Toggl.Daneel/Cells/Calendar/CalendarItemView.cs (hunk 1 above)
Comment (juanlaube) [reply Toggl.Daneel/Cells/Calendar/CalendarItemView.cs]: 👍

File: Toggl.Daneel/Cells/Calendar/CalendarItemView.cs (hunk 2 above)
Comment (juanlaube) [reply Toggl.Daneel/Cells/Calendar/CalendarItemView.cs]: Yes sir!

File: Toggl.Daneel/ViewSources/CalendarCollectionViewSource.cs (hunk 3 above)
Comment (juanlaube) [reply Toggl.Daneel/ViewSources/CalendarCollectionViewSource.cs]: I agree

File: Toggl.Daneel/Views/Calendar/CalendarCollectionViewEditItemHelper.cs (hunk 4 above)
Comment (juanlaube) [reply Toggl.Daneel/Views/Calendar/CalendarCollectionViewEditItemHelper.cs]: 🔥

File: Toggl.Daneel/Views/Calendar/CalendarCollectionViewEditItemHelper.cs (hunk 5 above)
Comment (juanlaube) [reply Toggl.Daneel/Views/Calendar/CalendarCollectionViewEditItemHelper.cs]: 🔥

File: Toggl.Foundation.MvvmCross/ViewModels/Calendar/CalendarViewModel.cs (hunk 9 above)
Comment (juanlaube) [reply Toggl.Foundation.MvvmCross/ViewModels/Calendar/CalendarViewModel.cs]: Yeah, this seems to be indeed the de facto way, even mimics what's done in the EditTimeEntryViewModel.

File: Toggl.Foundation/Interactors/InteractorFactory.TimeEntry.cs (hunk 12 above)
Comment (juanlaube) [reply Toggl.Foundation/Interactors/InteractorFactory.TimeEntry.cs]: There was no interactor, the TimeEntryViewModel does something similar to this.

File: Toggl.Daneel/Views/Calendar/CalendarCollectionViewEditItemHelper.cs (hunk 6 above)
Comment (juanlaube) [reply Toggl.Daneel/Views/Calendar/CalendarCollectionViewEditItemHelper.cs]: The only way is to validate that the touch happens on the cell for the itemIndexPath, that's already being handled.

File: Toggl.Foundation/Interactors/InteractorFactory.TimeEntry.cs (hunk 12 above)
Comment (willsb) [reply Toggl.Foundation/Interactors/InteractorFactory.TimeEntry.cs]: hmm... We might need to rethink this interactor, I guess. But for the time being, that's good enough :)

File: Toggl.Daneel/Cells/Calendar/CalendarItemView.cs [hunk 13]
```diff
@@ -25,7 +26,15 @@ public sealed partial class CalendarItemView : ReactiveCollectionViewCell<Calend
         public CGRect TopDragTouchArea => TopDragIndicator.Frame.Inset(-20, -20);
         public CGRect BottomDragTouchArea => BottomDragIndicator.Frame.Inset(-20, -20);
 
-        public bool IsEditing { get; set; }
+        private bool isEditing;
+        public bool IsEditing {
```
Comment (willsb) [inline Toggl.Daneel/Cells/Calendar/CalendarItemView.cs]: 🇪🇬

File: Toggl.Daneel/Cells/Calendar/CalendarItemView.cs [hunk 14]
```diff
@@ -16,6 +16,7 @@ public sealed partial class CalendarItemView : ReactiveCollectionViewCell<Calend
     {
         private static readonly Dictionary<CalendarIconKind, UIImage> images;
 
+        private UIColor itemColor => MvxColor.ParseHexString(Item.Color).ToNativeColor();
```
Comment (willsb) [inline Toggl.Daneel/Cells/Calendar/CalendarItemView.cs]: This being a property kind of bothers. It should be a method

**Pull request merged**